labs/KeyWrapping/KeyWrapping.md

```diff
@@ -0,0 +1,1119 @@
+takes and returns 32-bit values, we have to shrink `x` from `a` bits
+to `32` bits, and widen the result up to `48` bits. And, since `h`
+always takes and returns 64-bit values, we have to widen `x` from `a`
+bits to `64` bits, and shrink the result back down to `48` bits To
+help us do this resizing work, we'll introduce `shrink` and `widen`
+functions.
+
+```
+widen : {a, b} (fin a, fin b) => [b] -> [a + b]
+widen a = 0 # a
+
+shrink : {a, b} (fin a, fin b) => [a + b] -> [b]
+shrink a = drop a
+```
+
+`widen` takes any sized input and prepends `0` or more `False`
+bits. `shrink` takes any sized input and removes `0` or more bits from
+the front. Using these two functions, we can fix our `f` from above:
+
+```
+f : {a} (32 <= a, a <= 64) => [a] -> [48]
+f x = if `a <= 0x30 then
+        widen (g (shrink x))
+      else
+        shrink (h (widen x))
+```
+
+And here we test that `f` correctly calls `g` and `h` (which increment
+and decrement by 1, respectively).
+
+```sh
+labs::KeyWrapping::KeyWrapping> f (10 : [37])
+11
+labs::KeyWrapping::KeyWrapping> f (10 : [53])
+9
+```
+
+
+## KWP-AE Top Level Function
+
+With those two considerations firmly under our belt, we can now tackle
+`KWP-AE`.
+
+**EXERCISE**: Study `Algorithm 5` from the standard and complete the
+  definition of `KWPAE` below by filling in the function skeleton
+  provided with appropriate logic. Use the `KWPAEPad` function from
+  above to create `S`. Use the `shrink` and `widen` functions to
+  assist in resizing `S` and the function outputs on the `then` and
+  `else` branches of line 5.
+  
+*Hint:* You'll notice that we needed to pull in the type variable `n`
+and type constraints from `W` and relate `n` and `l` (the type of both
+`S` and `C`). It may also be necessary to tell `W` that our type
+variable `n` is the same type variable that it uses. This can be
+achieved by calling `W` like so: ```W`{n=n}```.
+
+```
+KWPAE :
+    {k, l, n}            // k is [len(P)/8], Algorithm 5
+    ( 1 <= k, k < 2^^32  // Bounds on the number of octets of P, from Table 1
+    , l == 32 + 32 + k*8 + k*8 %^ 64  // The type of S and C
+    , 64*n == max 192 l               // Here we relate n and l
+    ) =>
+    ([128] -> [128]) -> [k][8] -> [l]
+KWPAE CIPHk P = C
+  where
+    S = undefined : [l]
+    C = if (`k : [32]) <= 8 then
+          undefined : [l]
+        else
+          undefined : [l]
+```
+
+Feel free to use the provided `KWPAETests` property to check your
+work.
+
+
+# A Formal Specification of `KWP-AD`
+
+`KWP-AD` is the authenticated-decryption function and makes use of our
+previously defined `W'`.
+
+**EXERCISE**: Study `Algorithm 6` from the standard and complete the
+  definition of `KWPAD` below by filling in the function skeleton
+  provided with appropriate logic. We suggest splitting the algorithm
+  into two again, so we're providing a skeleton for `KWPADUnpad` (which
+  is roughly the inverse of `KWPAEPad`) and `KWPAD`.
+
+```
```

Comment:
Would it be helpful to hint that `FAIL` does not need to be written as an if/then/else even though that is the language used in the spec?